sbi/inference/posteriors/mcmc_posterior.py

```diff
@@ -352,8 +352,90 @@ def sample(
+        return samples
+
+    def sample_batched(
+        self,
+        sample_shape: Shape,
+        x: Tensor,
+        method: Optional[str] = None,
+        thin: Optional[int] = None,
+        warmup_steps: Optional[int] = None,
+        num_chains: Optional[int] = None,
+        init_strategy: Optional[str] = None,
+        init_strategy_parameters: Optional[Dict[str, Any]] = None,
+        num_workers: Optional[int] = None,
+        mp_context: Optional[str] = None,
+        show_progress_bars: bool = True,
+    ) -> Tensor:
+        r"""Given a batch of observations [x_1, ..., x_B] this function samples from
+        posteriors $p(\theta|x_1)$, ... ,$p(\theta|x_B)$, in a batched (i.e. vectorized)
+        manner.
+
+        Check the `__init__()` method for a description of all arguments as well as
+        their default values.
+
+        Args:
+            sample_shape: Desired shape of samples that are drawn from the posterior
+                given every observation.
+            x: A batch of observations, of shape `(batch_dim, event_shape_x)`.
+                `batch_dim` corresponds to the number of observations to be drawn.
+            show_progress_bars: Whether to show sampling progress monitor.
+
+        Returns:
+            Samples from the posteriors of shape (*sample_shape, B, *input_shape)
+        """
+        batch_size = x.shape[0]
+        self.potential_fn.set_x(x)
+
+        # Replace arguments that were not passed with their default.
+        method = self.method if method is None else method
+        thin = self.thin if thin is None else thin
+        warmup_steps = self.warmup_steps if warmup_steps is None else warmup_steps
+        num_chains = self.num_chains if num_chains is None else num_chains
+        init_strategy = self.init_strategy if init_strategy is None else init_strategy
+        num_workers = self.num_workers if num_workers is None else num_workers
+        mp_context = self.mp_context if mp_context is None else mp_context
+        init_strategy_parameters = (
+            self.init_strategy_parameters
+            if init_strategy_parameters is None
+            else init_strategy_parameters
+        )
+        self.potential_ = self._prepare_potential(method)  # type: ignore
+
+        # For each observation in the batch, we have num_chains independent chains.
+        num_chains_extended = batch_size * num_chains
+        initial_params = self._get_initial_params(
+            init_strategy,  # type: ignore
+            num_chains_extended,  # type: ignore
+            num_workers,
+            show_progress_bars,
+            **init_strategy_parameters,
+        )
+        # We need num_samples from each posterior in the batch
+        num_samples = torch.Size(sample_shape).numel() * batch_size
+
+        assert (
+            method == "slice_np_vectorized"
+        ), "Batched sampling only supported for vectorized samplers!"
```

Comment:
this check should happen at the top of the method because we might have spent some time finding initial params already, which would be in vein then, no?

Reply:
also, could we do the batched sampling for the other samplers as well? The speed-up would not be so much, but if one is using multiple CPU cores for different chains, one could still get a speed up, no?

Reply:
Yeah, one could also do parallel sampling per observation instead of parallel sampling per chain (maybe also both together).

Not sure if it is so easy to implement (but I will try for a bit).

Reply:
Not sure I understand what you mean by sampling per observation. My thought was to distribute the observations across chains, e.g., if we 10 observations, and originally 5 chains, we call the sampler with 50 chains and distribute the observations across them accordingly. For the SliceSerial that should be very similar to the approach we currently have for SliceVectorized, no? 
For Pyro and PyMC I am not sure. 

But anyway, if it is too much of a stretch now I think it's great already that we have the sampling for the vectorized slide sampler! 👏

Reply:
Yeah, I think we mean the same. The difference is what we do now is that the potential_fn must change for different chains (as we have different x_os). Not sure if this works out-of the box (but could be that simple, will try).